File: aat-gtk/src/main/java/ch/bailu/aat_gtk/lib/FileDialog.java
package ch.bailu.aat_gtk.lib;

import java.util.HashMap;
import java.util.Map;

import ch.bailu.gtk.gio.File;
import ch.bailu.gtk.gtk.FileChooser;
import ch.bailu.gtk.gtk.FileChooserAction;
import ch.bailu.gtk.gtk.FileChooserDialog;
import ch.bailu.gtk.gtk.FileFilter;
import ch.bailu.gtk.gtk.ResponseType;
import ch.bailu.gtk.gtk.Window;
import ch.bailu.gtk.type.Str;

public class FileDialog {

    private String title = "";
    private String cancel = "Cancel";
    private String ok = "OK";
    private String initialPath = "";
    private OnResponse onResponseCallback = path -> {};
    private Map<String, FileFilter> filterMap = new HashMap<>();

    private int action = FileChooserAction.OPEN;

    public interface OnResponse {
        void onResponse(String path);
    }

    public FileDialog title(String label) {
        this.title = label;
        return this;
    }

    public FileDialog ok(String label) {
        this.ok = label;
        return this;
    }

    public FileDialog cancel(String label) {
        this.cancel = label;
        return this;
    }

    public FileDialog open() {
        this.action = FileChooserAction.OPEN;
        return this;
    }

    public FileDialog save() {
        this.action = FileChooserAction.SAVE;
        return this;
    }

    public FileDialog mimeType(String name, String mimeType) {
        if (!filterMap.containsKey(name)) {
            var filter = new FileFilter();
            filter.setName(name);
            filterMap.put(name, filter);
        }
        filterMap.get(name).addMimeType(mimeType);
        return this;
    }

    public FileDialog pattern(String name, String pattern) {
        if (!filterMap.containsKey(name)) {
            var filter = new FileFilter();
            filter.setName(name);
            filterMap.put(name, filter);
        }
        filterMap.get(name).addPattern(pattern);
        return this;
    }

    public FileDialog onResponse(OnResponse response) {
        this.onResponseCallback = response;
        return this;
    }

    public FileDialog selectFolder() {
        this.action = FileChooserAction.SELECT_FOLDER;
        return this;
    }

    public FileDialog path(String path) {
        this.initialPath = path;
        return this;
    }

    public void show(Window window) {
        var dialog = new FileChooserDialog(title, window, action, null);

        var initialPath = new Str(this.initialPath);

        setInitialPath(dialog, initialPath);
        var chooser = new FileChooser(dialog.cast());
        filterMap.keySet().forEach((key) -> chooser.addFilter(filterMap.get(key)));

        dialog.addButton(cancel, ResponseType.CANCEL);
        dialog.addButton(ok, ResponseType.OK);
                dialog.onDestroy(() -> {
            dialog.disconnectSignals();
            initialPath.destroy();
        });
        dialog.onResponse((response) -> {
            if (response == ResponseType.OK) {
                var path = getPath(dialog);
                onResponseCallback.onResponse(path);
            }
            dialog.close();
        });
        dialog.show();
    }

    private static void setInitialPath(FileChooserDialog dialog, Str initialPath) {
        if (initialPath.getSize() > 1) {
            try {
                var path = File.newForPath(initialPath);
                new FileChooser(dialog.cast()).setFile(path);
                unref(path);

            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
        }
    }

    private static String getPath(FileChooserDialog dialog) {
        var chooser = new FileChooser(dialog.cast());
        var file = chooser.getFile();
        var path = file.getPath();
        var result = path.toString();

        unref(file);
        path.destroy();

        return result;
    }

    private static void unref(File file) {
        new ch.bailu.gtk.gobject.Object(file.cast()).unref();
    }
}

File: aat-gtk/src/main/kotlin/ch/bailu/aat_gtk/view/dialog/FileDialog.kt
package ch.bailu.aat_gtk.view.dialog

import ch.bailu.aat_lib.resources.Res
import ch.bailu.gtk.gtk.FileChooser
import ch.bailu.gtk.gtk.FileChooserAction
import ch.bailu.gtk.gtk.FileChooserDialogExtended
import ch.bailu.gtk.gtk.FileFilter
import ch.bailu.gtk.gtk.ResponseType
import ch.bailu.gtk.gtk.Window

class FileDialog {
    private var title = ""
    private val cancel = Res.str().cancel()
    private val ok = Res.str().ok()

    private var initialPath = ""
    private var onResponseCallback = { _: String -> }

    private val filterMap: MutableMap<String, FileFilter> = HashMap()

    private var action = FileChooserAction.OPEN

    fun title(label: String): FileDialog {
        title = label
        return this
    }

    fun open(): FileDialog {
        action = FileChooserAction.OPEN
        return this
    }

    fun save(): FileDialog {
        action = FileChooserAction.SAVE
        return this
    }

    fun mimeType(name: String, mimeType: String?): FileDialog {
        if (!filterMap.containsKey(name)) {
            val filter = FileFilter()
            filter.setName(name)
            filterMap[name] = filter
        }
        filterMap[name]?.apply { addMimeType(mimeType) }
        return this
    }

    fun pattern(name: String, pattern: String?): FileDialog {
        if (!filterMap.containsKey(name)) {
            val filter = FileFilter()
            filter.setName(name)
            filterMap[name] = filter
        }
        filterMap[name]!!.addPattern(pattern)
        return this
    }

    fun onResponse(onResponse: (String)->Unit): FileDialog {
        onResponseCallback = onResponse
        return this
    }

    fun selectFolder(): FileDialog {
        action = FileChooserAction.SELECT_FOLDER
        return this
    }

    fun path(path: String): FileDialog {
        initialPath = path
        return this
    }

    fun show(window: Window) {
        val dialog = FileChooserDialogExtended(title, window, action)
        dialog.path = initialPath

        val chooser = FileChooser(dialog.cast())
        filterMap.keys.forEach { key: String ->
            filterMap[key]?.apply { chooser.addFilter(this) }
        }

        dialog.addButton(cancel, ResponseType.CANCEL)
        dialog.addButton(ok, ResponseType.OK)
        dialog.onDestroy {
            dialog.disconnectSignals()
        }

        dialog.onResponse { response: Int ->
            if (response == ResponseType.OK) {
                onResponseCallback(dialog.path)
            }
            dialog.close()
        }
        dialog.show()
    }
}
